### DISCUSSION TITLE
Postgres connection performance implications

### DISCUSSION THREAD

**Discussion by notramo** [geldata/gel#6245] [Q&A]
I know that EdgeDB accepts read-only Postgres connection on the same port.
As the kernel can not decide if it's an EdgeDB  or Postgres client, I guess, there is some proxy running, which forwards Postgres connections as-is.
I have 2 questions regarding this:
- Are there any benchmarks that compare EdgeDB proxied Postgres connection with raw Postgres connection?
- If it's significantly slower, is it better to connect to the underlying Postgres instance directly for running high number of queries? Does it involve any pitfalls? (E.g. if EdgeDB tracks the number of connections it maintains to the database, compares them to the maximum configured Postgres connection pool, and gets confused by getting connection errors because the pool being used by other clients?)

**Comment by raddevon** [comment]
@aljazerzen or @fantix might be able to address your first question. I'm not aware of any benchmarks, but they may be able to speak to performance.

As for your second question, you will not see a familiar schema if you connect to the underlying Postgres, so I don't think the two alternatives you've proposed would be analogous. The schema in the underlying Postgres isn't meant to be read and manipulated directly.

**Reply by notramo** [reply]
After grokking through the Postgres contents, it seems clean enough to use raw SQL for some performance-critical read-only queries (of course, taking great care when schema is modified in .esdl file).

**Comment by elprans** [comment]
> Are there any benchmarks that compare EdgeDB proxied Postgres connection with raw Postgres connection?

Yes, [over here](https://github.com/edgedb/imdbench)

**Reply by notramo** [reply]
It's comparing EdgeQL with SQL, but I couldn't find EdgeDB-proxied `psql` vs. raw `psql`. No EdgeDB connection protocol, only Postgres.

`psql` (Postgres protocol) -> real PostgreSQL server
vs.
`psql` (Postgres protocol) -> EdgeDB fake Postgres -> PostgreSQL server

What I'm interested in is the EdgeDB proxy overhead when proxying Postgres protocol.

**Reply by notramo** [reply]
Also, is the proxy multithreaded?

**Reply by elprans** [reply]
The proxy is single-threaded, but is async.  There is no formal benchmark yet, but the overhead should be minimal.

**Reply by notramo** [reply]
Does `async` mean it's written in Python? Then I guess it's better to connect to the Postgres server directly, without the proxy, given that only the most performance-critical queries are ran through Postgres connection. Are there any pitfalls when connecting directly to Postgres compared with when connecting though the proxy?

**Reply by elprans** [reply]
> Does `async` mean it's written in Python? Then I guess it's better to connect to the Postgres server directly. 

It's written in Cython, effectively a C Python extension.  If you're still concerned about the performance of async code written this way, look at [asyncpg's performance](https://github.com/MagicStack/asyncpg) (which we also wrote).

> Are there any pitfalls when connecting directly to Postgres compared with when connecting though the proxy?

That is not a supported way of operating EdgeDB.  The underlying Postgres schema and data are owned by EdgeDB and are not intended for direct consumption.
